### PR TITLE
[browser][MT] fix library tests, remove MonoWasmBuildVariant=perftrace

### PR DESCRIPTION
**Pull request by pavelsavara**
Multi-threading tests:
- fix the msbuild scripts to detect MT build when building library tests and set 
- disable V8 and NodeJS for threading tests
- removed perftrace runtime flavor
- removed MonoWasmBuildVariant==perftrace
- made IsThreadingSupported respect build flavor via IsBrowserThreadingSupported and --use-threads
- removed most of event-pipe wasm sample
- removed DISABLE_WASM_USER_THREADS and MonoWasmThreadsNoUser
- fixed mono_wasm_execute_timer for WS without Main
- added COOP into SimpleServer
- fixed problem with SharedArrayBuffer in WS

This PR is not fixing tests which are actually broken, they are on optional CI pipeline, which doesn't fail the build.

### CODE REVIEW

**Comment by pavelsavara** [comment]
At least some lib tests are passing on MT, for example 
[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-87549-merge-32909bb61e6442c6a6/WasmTestOnBrowser-Common.Tests/1/console.879db6f0.log?helixlogtype=result)
[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-87549-merge-32909bb61e6442c6a6/WasmTestOnBrowser-Microsoft.Bcl.TimeProvider.Tests/1/console.5cdda350.log?helixlogtype=result)

V8 doesnn't have threads and EventTarget class, we should run all tests on the browser.
NodeJS is failing with https://github.com/dotnet/runtime/issues/87556 and we should disable it for now.

File: eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml [hunk 1]
```diff
@@ -137,9 +137,9 @@ jobs:
       # NOTE - Since threading is experimental, we don't want to block mainline work
       shouldContinueOnError: true
       scenarios:
-        - normal
+        #- normal
```
Comment (kg) [inline eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml:140]: Maybe add a comment explaining why these scenarios are disabled that mentions when we should re-enable them

File: eng/testing/tests.browser.targets [hunk 2]
```diff
@@ -96,6 +96,7 @@
     <_AppArgs Condition="'$(WasmTestAppArgs)' != ''">$(_AppArgs) $(WasmTestAppArgs)</_AppArgs>
 
     <WasmXHarnessMonoArgs Condition="'$(XunitShowProgress)' == 'true'">$(WasmXHarnessMonoArgs) --setenv=XHARNESS_LOG_TEST_START=1</WasmXHarnessMonoArgs>
+    <WasmXHarnessMonoArgs Condition="'$(WasmEnableThreads)' == 'true' or '$(WasmEnablePerfTracing)' == 'true' or '$(WasmEnablePerfTracing)' == 'true' or '$(MonoWasmBuildVariant)' == 'multithread'">$(WasmXHarnessMonoArgs) --no-memory-snapshot</WasmXHarnessMonoArgs>
```
Comment (kg) [inline eng/testing/tests.browser.targets:99]: This chain of 'or's is kind of unwieldy, it might be better to have a synthesized property like
```xml
<WasmSnapshotsAllowed Condition="'$(WasmEnableThreads)' == 'true'">false</WasmSnapshotsAllowed>
...
<WasmSnapshotsAllowed Condition="'$(WasmSnapshotsAllowed)' == ''">true</WasmSnapshotsAllowed>
<WasmXHarnessMonoArgs Condition="'$(WasmSnapshotsAllowed' == 'true'">$(WasmXHarnessMonoArgs) --no-memory-snapshot</WasmXHarnessMonoArgs>
```

File: src/mono/wasm/build/WasmApp.targets [hunk 3]
```diff
@@ -135,6 +135,8 @@
 
     <!-- by default, package assemblies as webcil -->
     <WasmEnableWebcil Condition="'$(WasmEnableWebcil)' == ''">true</WasmEnableWebcil>
+
+    <WasmEnableThreads Condition="'$(WasmEnablePerfTracing)' == 'true' or '$(MonoWasmBuildVariant)' == 'multithread'">true</WasmEnableThreads>
```
Comment (kg) [inline src/mono/wasm/build/WasmApp.targets:139]: I dislike this a lot, it would be better if enabling perf tracing without enabling threads caused the build to fail with an `<Error>`. Otherwise you will get surprising differences in behavior (no jiterpreter, no snapshots) by turning on a feature that seems totally unrelated

Comment (pavelsavara) [reply src/mono/wasm/build/WasmApp.targets:139]: @lambdageek @lewing let's remove "perf tracing" flavor. I think we don't have bandwidth to support it.

Comment (kg) [reply src/mono/wasm/build/WasmApp.targets:139]: Could we make it automatically enabled whenever threads are on? is it expensive?

**Review by kg — APPROVED** [review]

File: eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml (hunk 1 above)
Comment (radical) [reply eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml:140]: And add this to wherever mt work is being tracked.

Comment (pavelsavara) [reply eng/pipelines/extra-platforms/runtime-extra-platforms-wasm.yml:140]: Added item for NodeJS to https://github.com/dotnet/runtime/issues/85592
Removed v8 version because that doesn't have threads at all AFAIK

File: eng/testing/tests.browser.targets (hunk 2 above)
Comment (pavelsavara) [reply eng/testing/tests.browser.targets]: simplified by removing perftracing flavor

File: src/mono/wasm/build/WasmApp.targets (hunk 3 above)
Comment (pavelsavara) [reply src/mono/wasm/build/WasmApp.targets]: simplified by removing perftracing flavor

Comment (lambdageek) [reply src/mono/wasm/build/WasmApp.targets]: I think we can kill the `MonoWasmBuildVariant=perftrace` configuration (I don't htink it was seeing much use, and it's kind of strange).

I think we should keep the diagnostic server and event pipe support if someone is building the multi-threaded runtime.   (ie: if `MonoWasmBuildVariant=true` then we should enable `FeaturePerfTracing` when building the multi-threaded CoreLib).

`WasmEnablePerfTracing` is a user-visible toggle. I think the way things work today is that it is exclusive with `WasmEnableThreads` - a user project can set one or the other, but not both.  I think if a user project sets perftracing to true or it enables threads, the eventpipe and diagnostic server code is included in the app.  
I'm not sure if we should just ignore `WasmEnablePerfTracing` or just always error out if someone tries to set it to true and point them at `WasmEnableThreads` instead.

I think in a future .NET release it would be good to build a single-threaded version of eventpipe and the diagnostic server.

File: src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Current.Manifest/WorkloadManifest.targets.in [hunk 4]
```diff
@@ -176,11 +175,4 @@
     <Error Condition="'$(WasmBuildNative)' == 'true'"
       Text="WebAssembly workloads, required for native relinking, are only supported for projects targeting net6.0+ . Set %24(WasmBuildNative)=false to disable it." />
   </Target>
-
-  <Target Name="_ErrorDualWasmThreadProps"
-    Condition="('$(TargetsCurrent)' == 'true' or '$(TargetsNet7)' == 'true') and '$(RuntimeIdentifier)' == 'browser-wasm' and
-    '$(BrowserWorkloadDisabled)' != 'true' and '$(WasmEnableThreads)' == 'true' and '$(WasmEnablePerfTrace)' == 'true'"
-    BeforeTargets="Build">
-    <Error Text="WebAssembly workloads can only support one active threading mode at a time. Either set WasmEnableThreads or WasmEnablePerfTracing to true, but not both." />
```
Comment (lambdageek) [inline src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Current.Manifest/WorkloadManifest.targets.in:184]: I think we should keep this.

Also I think you have to keep the .NET 7 behavior

File: src/mono/sample/wasm/Directory.Build.props [hunk 5]
```diff
@@ -6,6 +6,7 @@
     <TargetOS>browser</TargetOS>
     <TargetArchitecture>wasm</TargetArchitecture>
     <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
+    <MonoWasmBuildVariant Condition="'$(WasmEnableThreads)' == 'true'">multithread</MonoWasmBuildVariant>
```
Comment (lambdageek) [inline src/mono/sample/wasm/Directory.Build.props:9]: This doesn't make sense to me.  the sample apps are like user apps.  but `MonoWasmBuildVariant` is a property of how we build the runtime.  By the time any of this code runs, the runtime is already built. How can setting it here have any effect?

File: src/mono/sample/wasm/browser-eventpipe/Program.cs [hunk 6]
```diff
@@ -1,162 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
Comment (lambdageek) [inline src/mono/sample/wasm/browser-eventpipe/Program.cs:1]: Why remove this? it should work with the threaded runtime

**Review by lambdageek — APPROVED** [review]

File: src/mono/sample/wasm/browser-eventpipe/Program.cs (hunk 6 above)
Comment (pavelsavara) [reply src/mono/sample/wasm/browser-eventpipe/Program.cs:1]: reverted

**Comment by SamMonoRT** [comment]
@pavelsavara @radical  - do we need any corresponding changes in any performance pipelines/lanes ?

File: src/mono/sample/wasm/browser-eventpipe/README.md [hunk 7]
```diff
@@ -1,10 +1,10 @@
 [see also](../../../wasm/runtime/diagnostics/README.md)
 
-To be able to run this sample you need to build the runtime with `/p:WasmEnablePerfTracing=true` and use Chrome browser
+To be able to run this sample you need to build the runtime with `/p:FeatureWasmThreads=true` and use Chrome browser
```
Comment (maraf) [inline src/mono/sample/wasm/browser-eventpipe/README.md:3]: Shouldn't it be `WasmEnableThreads=true`?

File: src/mono/wasm/test-main.js [hunk 8]
```diff
@@ -136,6 +136,8 @@ function processArguments(incomingArguments, runArgs) {
             runArgs.forwardConsole = false;
         } else if (currentArg == "--no-memory-snapshot") {
             runArgs.memorySnapshot = false;
+        } else if (currentArg == "--use-threads") {
```
Comment (maraf) [inline src/mono/wasm/test-main.js:139]: Shouldn't we reuse `--setenv` here?

Comment (pavelsavara) [reply src/mono/wasm/test-main.js:139]: much better, thanks

File: src/mono/sample/wasm/browser-eventpipe/README.md (hunk 7 above)
Comment (pavelsavara) [reply src/mono/sample/wasm/browser-eventpipe/README.md:3]: It should be `/p:MonoWasmBuildVariant=multithread` thanks!

**Review by maraf — APPROVED** [review]

**Comment by pavelsavara** [comment]
Here are MT test failures  this uncovered

Timeout probably because of the sample setup.
[Wasm.Browser.EventPipe.Sample](https://helix.dot.net/api/2019-06-17/jobs/78e38870-6381-4adb-b2d7-b11dd4d6025f/workitems/Wasm.Browser.EventPipe.Sample/console)

Maybe we need bigger thread pool or something is eating/leaking the threads.
`ThreadStartException: Thread failed to start.` 
[System.Threading.Tasks.Tests](https://helix.dot.net/api/2019-06-17/jobs/78e38870-6381-4adb-b2d7-b11dd4d6025f/workitems/WasmTestOnBrowser-System.Threading.Tasks.Tests/console)
[System.Transactions.Local.Tests](https://helix.dot.net/api/2019-06-17/jobs/78e38870-6381-4adb-b2d7-b11dd4d6025f/workitems/WasmTestOnBrowser-System.Transactions.Local.Tests/console)
[System.Threading.Tests](https://helix.dot.net/api/2019-06-17/jobs/78e38870-6381-4adb-b2d7-b11dd4d6025f/workitems/WasmTestOnBrowser-System.Threading.Tests/console)
 

`System.Threading.Tests.TimerFiringTests.ManyTimers_EachTimerDoesFire [10:42:43] info: Not all timers fired, 10000 left of 10000`
[System.Threading.Timer.Tests](https://helix.dot.net/api/2019-06-17/jobs/78e38870-6381-4adb-b2d7-b11dd4d6025f/workitems/WasmTestOnBrowser-System.Threading.Timer.Tests/console)
